File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/map/overlay/overlay/NPathOverlay.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay

import androidx.annotation.ColorInt
import com.naver.maps.geometry.LatLng
import com.naver.maps.map.overlay.PathOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.overlay.AddableOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asBoolean
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asInt
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asList
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLatLng
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.info.NOverlayInfo
import dev.note11.flutter_naver_map.flutter_naver_map.model.map.overlay.NOverlayImage
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil

internal data class NPathOverlay(
    override val info: NOverlayInfo,
    val coords: List<LatLng>,
    val widthDp: Double,
    @ColorInt val color: Int,
    val outlineWidthDp: Double,
    @ColorInt val outlineColor: Int,
    @ColorInt val passedColor: Int,
    @ColorInt val passedOutlineColor: Int,
    val progress: Double,
    val patternImage: NOverlayImage?,
    val patternIntervalDp: Double,
    val isHideCollidedCaptions: Boolean,
    val isHideCollidedMarkers: Boolean,
    val isHideCollidedSymbols: Boolean,
) : AddableOverlay<PathOverlay>() {


    override fun createMapOverlay(): PathOverlay = applyAtRawOverlay(PathOverlay())

    override fun applyAtRawOverlay(overlay: PathOverlay) = overlay.also { g ->
        g.coords = coords
        g.width = DisplayUtil.dpToPx(widthDp)
        g.color = color
        g.outlineColor = outlineColor
        g.outlineWidth = DisplayUtil.dpToPx(outlineWidthDp)
        g.passedColor = passedColor
        g.passedOutlineColor = passedOutlineColor
        g.progress = progress
        patternImage?.applyToOverlay(g::setPatternImage)
        g.patternInterval = DisplayUtil.dpToPx(patternIntervalDp)
        g.isHideCollidedCaptions = isHideCollidedCaptions
        g.isHideCollidedMarkers = isHideCollidedMarkers
        g.isHideCollidedSymbols = isHideCollidedSymbols
    }

    companion object {
        fun fromMessageable(rawMap: Any): NPathOverlay = rawMap.asMap().let {
            NPathOverlay(
                info = NOverlayInfo.fromMessageable(it[infoName]!!),
                coords = it[coordsName]!!.asList { l -> l.asLatLng() },
                widthDp = it[widthName]!!.asDouble(),
                color = it[colorName]!!.asInt(),
                outlineColor = it[outlineColorName]!!.asInt(),
                outlineWidthDp = it[outlineWidthName]!!.asDouble(),
                passedColor = it[passedColorName]!!.asInt(),
                passedOutlineColor = it[passedOutlineColorName]!!.asInt(),
                progress = it[progressName]!!.asDouble(),
                patternImage = it[patternImageName]?.let(NOverlayImage::fromMessageable),
                patternIntervalDp = it[patternIntervalName]!!.asDouble(),
                isHideCollidedCaptions = it[isHideCollidedCaptionsName]!!.asBoolean(),
                isHideCollidedMarkers = it[isHideCollidedMarkersName]!!.asBoolean(),
                isHideCollidedSymbols = it[isHideCollidedSymbolsName]!!.asBoolean(),
            )
        }

        /*
            --- Messaging Name Define ---
        */
        private const val infoName = "info"
        const val coordsName = "coords"
        const val widthName = "width"
        const val colorName = "color"
        const val outlineWidthName = "outlineWidth"
        const val outlineColorName = "outlineColor"
        const val passedColorName = "passedColor"
        const val passedOutlineColorName = "passedOutlineColor"
        const val progressName = "progress"
        const val patternImageName = "patternImage"
        const val patternIntervalName = "patternInterval"
        const val isHideCollidedCaptionsName = "isHideCollidedCaptions"
        const val isHideCollidedMarkersName = "isHideCollidedMarkers"
        const val isHideCollidedSymbolsName = "isHideCollidedSymbols"
        const val boundsName = "bounds"
    }
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/model/map/overlay/overlay/NPathOverlay.swift
import NMapsMap

internal struct NPathOverlay: AddableOverlay {
    typealias OverlayType = NMFPath
    var overlayPayload: Dictionary<String, Any?> = [:]

    let info: NOverlayInfo
    let coords: Array<NMGLatLng>
    let width: CGFloat
    let color: UIColor
    let outlineWidth: CGFloat
    let outlineColor: UIColor
    let passedColor: UIColor
    let passedOutlineColor: UIColor
    let progress: Double
    let patternImage: NOverlayImage?
    let patternInterval: Double
    let isHideCollidedCaptions: Bool
    let isHideCollidedMarkers: Bool
    let isHideCollidedSymbols: Bool

    func createMapOverlay() -> OverlayType {
        let overlay = NMFPath()
        return applyAtRawOverlay(overlay)
    }
    
    func applyAtRawOverlay(_ overlay: NMFPath) -> NMFPath {
        overlay.path = NMGLineString(points: coords)
        overlay.width = width
        overlay.color = color
        overlay.outlineWidth = outlineWidth
        overlay.outlineColor = outlineColor
        overlay.passedColor = passedColor
        overlay.passedOutlineColor = passedOutlineColor
        overlay.progress = progress
        overlay.patternIcon = patternImage?.overlayImage
        overlay.patternInterval = UInt(round(patternInterval))
        overlay.isHideCollidedCaptions = isHideCollidedCaptions
        overlay.isHideCollidedMarkers = isHideCollidedMarkers
        overlay.isHideCollidedSymbols = isHideCollidedSymbols
        return overlay
    }

    static func fromMessageable(_ v: Any) -> NPathOverlay {
        let d = asDict(v)
        return NPathOverlay(
                info: NOverlayInfo.fromMessageable(d[infoName]!),
                coords: asArr(d[coordsName]!, elementCaster: asLatLng),
                width: asCGFloat(d[widthName]!),
                color: asUIColor(d[colorName]!),
                outlineWidth: asCGFloat(d[outlineWidthName]!),
                outlineColor: asUIColor(d[outlineColorName]!),
                passedColor: asUIColor(d[passedColorName]!),
                passedOutlineColor: asUIColor(d[passedOutlineColorName]!),
                progress: asDouble(d[progressName]!),
                patternImage: castOrNull(d[patternImageName], caster: NOverlayImage.fromMessageable),
                patternInterval: asDouble(d[patternIntervalName]!),
                isHideCollidedCaptions: asBool(d[isHideCollidedCaptionsName]!),
                isHideCollidedMarkers: asBool(d[isHideCollidedMarkersName]!),
                isHideCollidedSymbols: asBool(d[isHideCollidedSymbolsName]!)
        )
    }

    /*
        --- Messaging Name Define ---
    */
    private static let infoName = "info"
    static let coordsName = "coords"
    static let widthName = "width"
    static let colorName = "color"
    static let outlineWidthName = "outlineWidth"
    static let outlineColorName = "outlineColor"
    static let passedColorName = "passedColor"
    static let passedOutlineColorName = "passedOutlineColor"
    static let progressName = "progress"
    static let patternImageName = "patternImage"
    static let patternIntervalName = "patternInterval"
    static let isHideCollidedCaptionsName = "isHideCollidedCaptions"
    static let isHideCollidedMarkersName = "isHideCollidedMarkers"
    static let isHideCollidedSymbolsName = "isHideCollidedSymbols"
    static let boundsName = "bounds"
}
